tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/WithGlobals.scala

```diff
@@ -0,0 +1,95 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js tools             **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2017, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+package org.scalajs.core.tools.linker.backend.emitter
+
+import GlobalRefUtils.unionPreserveEmpty
+
+/** A monad that associates a set of global variable names to a value.
+ *
+ *  Definition:
+ *  {{{
+ *  WithGlobals.apply(a) === WithGlobals(a, ∅)
+ *
+ *  WithGlobals(v, w).flatMap(f) === WithGlobals(f(v).value, w ∪ f(v).globalVarNames)
+ *  }}}
+ *
+ *  Left identity:
+ *  {{{
+ *               WithGlobals.apply(a).flatMap(f)     =?= f(a)
+ *               --------------------
+ *           WithGlobals(a, ∅).flatMap(f)            =?= f(a)
+ *           ----------------------------
+ *  WithGlobals(f(a).value, ∅ ∪ f(a).globalVarNames) =?= f(a)
+ *                          -----------------------
+ *  WithGlobals(f(a).value,   f(a).globalVarNames  ) =?= f(a)
+ *  }}}
+ *
+ *  Right identity:
+ *  {{{
+ *  WithGlobals(v, w).flatMap(WithGlobals.apply)                                     =?= WithGlobals(v, w)
+ *  --------------------------------------------
+ *  WithGlobals(WithGlobals.apply(v).value, w ∪ WithGlobals.apply(v).globalVarNames) =?= WithGlobals(v, w)
+ *              --------------------------      -----------------------------------
+ *  WithGlobals(          v               , w ∪                 ∅                    =?= WithGlobals(v, w)
+ *  }}}
+ *
+ *  Associativity:
+ *  {{{
+ *  WithGlobals(v, w).flatMap(f).flatMap(g) =?= WithGlobals(v, w).flatMap(x => f(x).flatMap(g))
+ *  ------------
+ *  WithGlobals(f(v).value, w ∪ f(v).globalVarNames).flatMap(g) =?= ...
+ *  -----------------------------------------------------------
+ *  WithGlobals(g(f(v).value).value, (w ∪ f(v).globalVarNames) ∪ g(f(v).value).globalVarNames) =?= ...
+ *
+ *  ... =?= WithGlobals(v, w).flatMap(x => f(x).flatMap(g))
+ *          -----------------------------------------------
+ *  ... =?= WithGlobals((x => f(x).flatMap(g))(v).value, w ∪ (x => f(x).flatMap(g))(v).globalVarNames)
+ *                      -------------------------------      ----------------------------------------
+ *
+ *  Aside:
+ *    (x => f(x).flatMap(g))(v)
+ *    === f(v).flatMap(g)
+ *    === WithGlobals(g(f(v).value).value, f(v).globalVarNames ∪ g(f(v).value).globalVarNames)
+ *
+ *  ... =?= WithGlobals(g(f(v).value).value, w ∪ (f(v).globalVarNames ∪ g(f(v).value).globalVarNames))
```

Comment:
No. The story went like:
1. I'm writing these `val (tree1, globalVarNames1) = genSomething(); val (tree2, globalVarNames2) = genSomethingElse(); (combineSomehow(tree1, tree2), union(globalVarNames1, globalVarNames2))` everywhere. I need to fix this.
2. Hey, Scala has this cool syntax with for comprehensions. I can use it if I write a data structure with a `map` and a `flatMap`
3. Come up with the implementation of `WithGlobals`. Code becomes clean, me happy.
4. Oh, `WithGlobals` really looks like a monad. Let me document it as such.
5. But wait, if I *do* document it as a monad, but it's not, I'm going to be attacked by an army of cats.
6. The logical reaction would be: don't document it as a monad.
7. My reaction was: let's prove it's a monad.
8. That happened.